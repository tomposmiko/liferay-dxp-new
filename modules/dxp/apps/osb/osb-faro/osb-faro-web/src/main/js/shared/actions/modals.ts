import {Modal} from 'shared/types';

const actionTypes = Modal.actionTypes;
const modalTypes = Modal.modalTypes;

export {actionTypes, modalTypes};

export const close = () => ({
	type: actionTypes.CLOSE_MODAL
});

export const closeAll = () => ({
	type: actionTypes.CLOSE_ALL_MODALS
});

export const open = (
	type,
	props: {[key: string]: any} = {},
	options: {closeOnBlur?: boolean} = {}
) => {
	const {closeOnBlur = true} = options;

	return {
		payload: {
			closeOnBlur,
			props,
			type
		},
		type: actionTypes.OPEN_MODAL
	};
};
