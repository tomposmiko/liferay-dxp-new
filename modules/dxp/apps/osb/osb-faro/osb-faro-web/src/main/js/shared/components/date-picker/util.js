import moment from 'moment';
import {conformsTo, isNil} from 'lodash';

/**
 * Represents a time between two dates.
 * @typedef Range
 * @property {Moment} start - The start of the range.
 * @property {Moment} end - The end of the range.
 */

export function isAboveMaxRange(date, maxRange) {
	return (
		isRange(date) &&
		date.start &&
		date.end &&
		maxRange &&
		moment
			.duration({
				from: date.start,
				to: date.end
			})
			.asDays() > maxRange
	);
}

/**
 * A validation function that determines if a value is a valid
 * moment instance, allowing for null.
 * @param {Object} value - The value to test.
 * @returns {boolean}
 */
function isMomentOrNil(value) {
	return isNil(value) || moment.isMoment(value);
}

/**
 * A validation function that determines if the value is a valid
 * moment or range.
 * @param {Object} value - The value to test.
 * @returns {boolean}
 */
export function isDateOrRange(value) {
	return isNil(value) || moment.isMoment(value) || isRange(value);
}

/**
 * Determines if a date is in a {@link Range}.
 * @param {Range|Moment} dateOrRange - The range to check in.
 * @param {Moment} date - The date to check for.
 * @returns {boolean}
 */
export function isInRange(dateOrRange, date) {
	if (moment.isMoment(dateOrRange)) {
		return false;
	}

	const {end, start} = dateOrRange;

	return start && date.isAfter(start) && end && date.isBefore(end);
}

/**
 * A validation function that determines if a value is a valid {@link Range}.
 * @param {Object} value - The value to test.
 * @returns {boolean}
 */
export function isRange(value) {
	return conformsTo(value, {
		end: isMomentOrNil,
		start: isMomentOrNil
	});
}

/**
 * Adds a date to an existing {@link Range}.
 *
 * This function will try to be smart and first update the
 * start or end if either are null. If both exist, then it will return
 * a new range with the start set to the new date, and the end reset
 * to null.
 * @param {Range} range - The range to update.
 * @param {Moment} date - The date to add to the range.
 * @returns {Range}
 */
export function updateRange(range, date) {
	if (!range.start) {
		return {
			...range,
			start: date
		};
	}

	if (range.start && !range.end) {
		if (date.isSameOrAfter(range.start)) {
			return {
				...range,
				end: date
			};
		} else {
			return {
				end: range.start,
				start: date
			};
		}
	}

	if (date.isSame(range.start) && date.isSame(range.end)) {
		return range;
	}

	return {
		end: null,
		start: date
	};
}
