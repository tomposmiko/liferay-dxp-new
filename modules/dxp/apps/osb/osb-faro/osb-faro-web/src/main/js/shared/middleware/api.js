import sendRequest from 'shared/util/request';
import {get, isNil} from 'lodash';

export const CALL_API = 'CALL_API';

export function toAction(type, ...objs) {
	const action = Object.assign(...objs, {type});

	delete action.meta[CALL_API];

	return action;
}

export default () => next => action => {
	const request = get(action, ['meta', CALL_API]);

	if (isNil(request)) {
		return next(action);
	}

	const {data, requestFn, types} = request;

	const [requestType, successType, failureType] = types;

	next(toAction(requestType, action));

	const retVal = requestFn ? requestFn(data) : sendRequest(request);

	return retVal.then(
		payload => {
			next(
				toAction(successType, action, {
					meta: {
						...action.meta,
						schema: request.schema
					},
					payload
				})
			);

			return {payload};
		},
		error => {
			next(
				toAction(failureType, action, {
					error: true
				})
			);

			throw error;
		}
	);
};
