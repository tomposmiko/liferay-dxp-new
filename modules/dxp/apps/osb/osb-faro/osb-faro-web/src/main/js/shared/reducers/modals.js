import {actionTypes} from '../actions/modals';
import {createReducer} from 'redux-toolbox';
import {List, Map} from 'immutable';

export default createReducer(new List(), {
	[actionTypes.CLOSE_ALL_MODALS](state) {
		return state.clear();
	},

	[actionTypes.CLOSE_MODAL](state) {
		return state.pop();
	},

	[actionTypes.OPEN_MODAL](state, action) {
		const {closeOnBlur, props, type} = action.payload;

		return state.push(
			new Map({
				closeOnBlur,
				props: new Map(props),
				type
			})
		);
	}
});
