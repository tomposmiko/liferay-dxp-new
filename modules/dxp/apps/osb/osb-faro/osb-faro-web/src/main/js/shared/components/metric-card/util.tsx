import MetricValue from './MetricValue';
import React, {Fragment} from 'react';
import Trend from 'shared/components/Trend';
import {CHART_COLOR_NAMES} from 'shared/components/Chart';
import {get, last} from 'lodash';
import {
	getAxisFormatter,
	getDataFormatter,
	getIntervals,
	getMetricFormatter
} from 'shared/util/charts';
import {getIcon, getStatsColor} from 'shared/util/metrics';
import {INTERVAL_KEY_MAP} from 'shared/util/time';
import {MetricType} from './metrics';
import {toRounded} from 'shared/util/numbers';
import {toUnix} from 'shared/util/date';

const CHART_DATA_ID_1 = 'data_1';
const CHART_DATA_ID_2 = 'data_2';
const CHART_DATA_PREVIOUS = 'data_previous';
const METRIC_TOOLTIP_LABEL_MAP = {
	bounceRateMetric: Liferay.Language.get('avg-bounce')
};

const {
	martell: CHART_GREEN,
	martellL2: CHART_GREEN_L2,
	mormont: CHART_ORANGE,
	stark: CHART_BLUE,
	starkL2: CHART_BLUE_L2
} = CHART_COLOR_NAMES;

export const Icons = {
	negative: 'caret-bottom-l',
	neutral: undefined,
	positive: 'caret-top-l'
};

const PREVIOUS_PERIOD_VISITORS_COLOR = '#393A4A';

type TBuildTabs = (props: {
	activeItemIndex: number;
	items: {
		content: {
			details: {
				color: string;
				icon: string;
				label: string;
			};
			title: string;
			type: MetricType;
			value: string;
		};
	}[];
	onActiveItemIndexChange: (index: number) => void;
}) => {
	onClick: () => void | ((index: number) => void);
	secondaryInfo: React.ReactElement;
}[];

export const buildTabs: TBuildTabs = ({
	activeItemIndex,
	items,
	onActiveItemIndexChange
}) =>
	items.map(({content}, index) => {
		const {details, title, type, value} = content;
		const {color, icon, label} = details;

		return {
			onClick: () => {
				if (activeItemIndex !== index && !onActiveItemIndexChange)
					return () => {};

				return onActiveItemIndexChange(index);
			},
			secondaryInfo: (
				<span>
					<span className='primary-content'>
						<MetricValue type={type} value={value} />
					</span>

					{label && <Trend color={color} icon={icon} label={label} />}
				</span>
			),
			tabId: index,
			title
		};
	});

export const getMetricName = (activeItemIndex, metrics) =>
	metrics.map(({name}) => name)[activeItemIndex];

export const getActiveItem = (retVal, compareToPrevious) => {
	if (!retVal) {
		return {
			chartData: [],
			intervals: [],
			timeline: []
		};
	}

	const chartData = retVal.data.slice(0, -1);
	const timeline = last(retVal.data);

	if (!compareToPrevious && retVal.asymmetricComparison) {
		retVal = {
			...retVal,
			chartData: chartData.map(dataSet => ({
				...dataSet,
				data: dataSet.data.slice(1)
			})),
			intervals: retVal.intervals.slice(1),

			// @ts-ignore

			timeline: {data: timeline.data.slice(1), id: timeline.id}
		};

		if (retVal.compositeData) {
			const compositeDataKeys = Object.keys(retVal.compositeData);

			retVal = {
				...retVal,
				compositeData: compositeDataKeys.reduce((acc, val) => {
					acc = {
						...acc,
						[val]: retVal.compositeData[val].slice(1)
					};

					return acc;
				}, {})
			};
		}
	} else if (compareToPrevious && retVal.asymmetricComparison) {
		retVal = {
			...retVal,
			chartData: chartData.map(dataSet => ({
				...dataSet,
				data:
					dataSet.id !== 'data_previous'
						? [null, ...dataSet.data.slice(1)]
						: dataSet.data
			})),
			timeline
		};
	} else {
		retVal = {
			...retVal,
			chartData,
			timeline
		};
	}

	return retVal;
};

export const getPreviousValueFromCompositeData = (
	compositeData,
	dataName,
	dateKey
) => {
	const data = get(compositeData, dataName);

	if (data) {
		return data.find(val => toUnix(val.key) === dateKey).previousValue;
	}
};

export const getRegexType = (type: MetricType): RegExp => {
	if (type === MetricType.Ratings) {
		return /([/][0-9]+)/g;
	} else {
		return /([a-zA-Z%])+/g;
	}
};

export const formatValue = (
	value: string,
	regex: RegExp
): React.ReactElement[] => {
	const items = value.split(' ');

	return items.map((item, i) => {
		const [value, unit] = item.split(regex);

		return (
			<Fragment key={i}>
				{value}

				<span className='metric-value-letter'>{unit}</span>
			</Fragment>
		);
	});
};

export const getMetricCardTabsData = (result, metrics) =>
	metrics.map(({name, title, type}) => {
		const metricFormatter = getMetricFormatter(type);

		return {
			content: {
				details: {
					color: getStatsColor(
						result[name].trend.trendClassification
					),
					icon: getIcon(result[name].trend.percentage),
					label: `${toRounded(
						Math.abs(result[name].trend.percentage)
					)}%`
				},
				name,
				title,
				type,
				value: metricFormatter(result[name].value)
			}
		};
	});

export const getMetricsData = (
	result,
	metrics,
	rangeSelectors = {},
	chartDataMapFn = getMetricsChartData,
	interval = INTERVAL_KEY_MAP.day
) =>
	metrics.map(({compositeMetrics, name, title, tooltipTitle, type}) =>
		getMetricData({
			chartDataMapFn,
			compositeMetrics,
			interval,
			name,
			rangeSelectors,
			result,
			title,
			tooltipTitle,
			type
		})
	);

export const getMetricsChartData = ({
	histogram,
	name,
	title,
	tooltipTitle,
	type
}) => [
	{
		color: CHART_BLUE,
		data: getDataFormatter(type)(histogram.map(({value}) => value)),
		id: CHART_DATA_ID_1,
		name: tooltipTitle || METRIC_TOOLTIP_LABEL_MAP[name] || title,
		tooltipTitle
	},
	{
		color: CHART_BLUE_L2,
		data: getDataFormatter(type)(
			histogram.map(({previousValue}) => previousValue)
		),
		id: CHART_DATA_PREVIOUS,
		name: Liferay.Language.get('previous-period')
	},
	{
		data: histogram.map(({key}) => key),
		id: 'x'
	}
];

export const convertHistogramKeysToDate = ({
	key,
	previousValueKey,
	valueKey,
	...otherParams
}) => ({
	key: toUnix(key),
	previousValueKey: previousValueKey.split('/').map(toUnix),
	valueKey: valueKey.split('/').map(toUnix),
	...otherParams
});

export const getMetricData = ({
	chartDataMapFn = getMetricsChartData,
	compositeMetrics,
	interval = INTERVAL_KEY_MAP.day,
	name,
	rangeSelectors,
	result,
	title,
	tooltipTitle,
	type
}) => {
	const metricFormatter = getMetricFormatter(type);

	const histogram = result[name].histogram.metrics.map(
		convertHistogramKeysToDate
	);

	const compositeData = compositeMetrics
		? {
				compositeContent: compositeMetrics.reduce(
					(
						acc,
						{
							name: compositeMetricName,
							title: compositeMetricTitle,
							type: compositeMetricType
						}
					) => {
						acc[compositeMetricName] = {
							details: {
								color: getStatsColor(
									result[compositeMetricName].trend
										.trendClassification
								),
								icon: getIcon(
									result[compositeMetricName].trend.percentage
								),
								label: `${toRounded(
									Math.abs(
										result[compositeMetricName].trend
											.percentage
									)
								)}%`
							},
							name: compositeMetricName,
							title: compositeMetricTitle,
							type: compositeMetricType,
							value: metricFormatter(
								result[compositeMetricName].value
							)
						};

						return acc;
					},
					{}
				),
				compositeData: compositeMetrics.reduce(
					(acc, {name: compositeMetricName}) => {
						acc[compositeMetricName] =
							result[compositeMetricName].histogram.metrics;

						return acc;
					},
					{}
				)
		  }
		: {};

	const dateKeysIMap = new Map(
		histogram.map(({key, valueKey}) => [key, valueKey])
	);

	return {
		...compositeData,
		asymmetricComparison: result[name].histogram.asymmetricComparison,
		content: {
			details: {
				color: getStatsColor(result[name].trend.trendClassification),
				icon: getIcon(result[name].trend.percentage),
				label: `${toRounded(Math.abs(result[name].trend.percentage))}%`
			},
			name,
			title,
			type,
			value: metricFormatter(result[name].value)
		},
		data: chartDataMapFn({
			...compositeData,
			histogram,
			name,
			title,
			tooltipTitle,
			type
		}),
		dateKeysIMap,
		format: getAxisFormatter(type),
		intervals: getIntervals(
			rangeSelectors.rangeKey,
			histogram.map(({key}) => key),
			interval,
			dateKeysIMap
		),
		prevDateKeysIMap: new Map(
			histogram.map(({key, previousValueKey}) => [key, previousValueKey])
		)
	};
};

export const getSiteMetricsChartData = ({
	compositeData,
	histogram,
	name,
	title,
	tooltipTitle,
	type
}) =>
	name === 'visitorsMetric'
		? [
				{
					color: CHART_BLUE,
					data: getDataFormatter(type)(
						compositeData.knownVisitorsMetric.map(
							({value}) => value
						)
					),
					dataName: 'knownVisitorsMetric',
					id: CHART_DATA_ID_1,
					name: Liferay.Language.get('known-visitors'),
					tooltipTitle: Liferay.Language.get('known'),
					type: 'bar'
				},
				{
					color: CHART_ORANGE,
					data: getDataFormatter(type)(
						compositeData.anonymousVisitorsMetric.map(
							({value}) => value
						)
					),
					dataName: 'anonymousVisitorsMetric',
					id: CHART_DATA_ID_2,
					name: Liferay.Language.get('anonymous-visitors'),
					tooltipTitle: Liferay.Language.get('anonymous'),
					type: 'bar'
				},
				{
					color: PREVIOUS_PERIOD_VISITORS_COLOR,
					data: getDataFormatter(type)(
						histogram.map(({previousValue}) => previousValue)
					),
					id: CHART_DATA_PREVIOUS,
					name: Liferay.Language.get('previous-period'),
					type: 'line'
				},
				{
					data: histogram.map(({key}) => key),
					id: 'x'
				}
		  ]
		: getMetricsChartData({histogram, name, title, tooltipTitle, type}).map(
				data =>
					[CHART_DATA_ID_1, CHART_DATA_PREVIOUS].includes(data.id)
						? {
								...data,
								color:
									data.id === CHART_DATA_PREVIOUS
										? CHART_GREEN_L2
										: CHART_GREEN
						  }
						: data
		  );
