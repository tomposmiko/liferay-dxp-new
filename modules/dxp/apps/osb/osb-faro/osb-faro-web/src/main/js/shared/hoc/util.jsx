import ErrorDisplay from 'shared/components/ErrorDisplay';
import ErrorPage from 'shared/pages/ErrorPage';
import LoadingPage from 'shared/pages/Loading';
import NoResultsDisplay from 'shared/components/NoResultsDisplay';
import React from 'react';
import Spinner from 'shared/components/Spinner';
import {compose} from 'redux';
import {get, omit} from 'lodash';
import {sub} from 'shared/util/lang';
import {toRoute} from 'shared/util/router';

/**
 * HOC for ErrorDisplay.
 * @param {Object} - Options object to pass as props to ErrorDisplay
 * @returns {Function} Returns the ErrorDisplay or Wrapped Component.
 */
export const withError =
	(options = {}) =>
	Component =>
	({error, errorProps = {}, pageDisplay = true, refetch, ...otherProps}) => {
		const otherOptions = omit(options, 'page');

		if (error) {
			return get(options, 'page', pageDisplay) ? (
				<ErrorPage {...errorProps} {...otherOptions} />
			) : (
				<ErrorDisplay
					onReload={refetch}
					spacer
					{...errorProps}
					{...otherOptions}
				/>
			);
		}

		return <Component refetch={refetch} {...otherProps} />;
	};

/**
 * HOC for NoResultsDisplay.
 * @param {Object} - Options object to pass as props to NoResultsDisplay.
 * @returns {Function} Returns the NoResultsDisplay or WrappedComponent.
 */
export const withEmpty =
	(options = {}) =>
	Component =>
	({
		data,
		error,
		loading,
		noResultsRenderer,
		noResultsRendererProps,
		total,
		...props
	}) => {
		if (((data && data.total === 0) || total === 0) && !loading && !error) {
			if (noResultsRenderer) {
				const NoResults = noResultsRenderer;

				return <NoResults />;
			}

			return (
				<NoResultsDisplay {...options} {...noResultsRendererProps} />
			);
		}

		return (
			<Component
				data={data}
				error={error}
				loading={loading}
				total={total}
				{...props}
			/>
		);
	};

/**
 * HOC for Loading display.
 * @param {Object} - Options object to pass as props to Loading component.
 * @returns {Function} Returns the Loading or WrappedComponent.
 */
export const withLoading =
	(options = {}) =>
	Component =>
	({
		alignCenter = false,
		className,
		data,
		fadeIn = true,
		inline = false,
		loading,
		pageDisplay = true,
		...otherProps
	}) => {
		if (loading) {
			return get(options, 'page', pageDisplay) ? (
				<LoadingPage
					className={className}
					fadeIn={get(options, 'fadeIn', fadeIn)}
					key='LOADING'
				/>
			) : (
				<Spinner
					alignCenter={get(options, 'alignCenter', alignCenter)}
					className={className}
					inline={get(options, 'inline', inline)}
					key='SPINNER'
					spacer={!get(options, 'inline', inline)}
				/>
			);
		}

		return <Component className={className} data={data} {...otherProps} />;
	};

export const withNull =
	(key, errorProps = {}) =>
	Component =>
	props => {
		const {entityType = Liferay.Language.get('page'), linkRoute} =
			errorProps;

		if (key && !props[key]) {
			return (
				<ErrorPage
					{...props}
					href={toRoute(linkRoute, props.router.params)}
					linkLabel={sub(Liferay.Language.get('go-to-x'), [
						entityType
					])}
					message={sub(
						Liferay.Language.get(
							'the-x-you-are-looking-for-does-not-exist'
						),
						[entityType.toLowerCase()]
					)}
					subtitle={sub(Liferay.Language.get('x-not-found'), [
						entityType
					])}
					title={Liferay.Language.get('404')}
				/>
			);
		}

		return <Component {...props} />;
	};

/**
 * HOC for displaying results.
 */
export const SafeResults = compose(
	withLoading(),
	withError()
)(({children, data}) => children(data));

export const WrapSafeResults = compose(
	withLoading(),
	withError()
)(({children}) => children);
