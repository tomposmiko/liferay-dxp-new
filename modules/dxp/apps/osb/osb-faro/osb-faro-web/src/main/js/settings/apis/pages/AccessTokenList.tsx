import * as API from 'shared/api';
import BasePage from 'settings/components/BasePage';
import Button from 'shared/components/Button';
import Card from 'shared/components/Card';
import CopyButton from 'shared/components/CopyButton';
import moment from 'moment';
import Promise from 'metal-promise';
import React, {useState} from 'react';
import Table from 'shared/components/table';
import TokenCell from '../components/TokenCell';
import URLConstants from 'shared/util/url-constants';
import {AccessToken} from '../types';
import {addAlert} from 'shared/actions/alerts';
import {Alert} from 'shared/types';
import {ApisPath} from 'shared/util/url-constants';
import {close, modalTypes, open} from 'shared/actions/modals';
import {compose} from 'redux';
import {connect, ConnectedProps} from 'react-redux';
import {formatDateToTimeZone, getDateNow} from 'shared/util/date';
import {RootState} from 'shared/store';
import {sub} from 'shared/util/lang';
import {
	withAdminPermission,
	withError,
	withLoading,
	withQuery
} from 'shared/hoc';

export const isExpired = (expirationDate: string) =>
	moment.utc(expirationDate).isSameOrBefore(getDateNow());

const DATE_FORMAT = 'MMM DD, YYYY';

const connector = connect(
	(store: RootState, {groupId}: {groupId: string}) => ({
		timeZoneId: store.getIn([
			'projects',
			groupId,
			'data',
			'timeZone',
			'timeZoneId'
		])
	}),
	{addAlert, close, open}
);

type PropsFromRedux = ConnectedProps<typeof connector>;

const TokenList: React.FC<
	{
		groupId: string;
		refetch: () => Promise<any>;
		tokens: AccessToken[];
	} & PropsFromRedux
> = ({addAlert, close, groupId, open, refetch, timeZoneId, tokens}) => {
	const [loading, setLoading] = useState(false);

	const handleError = () => {
		setLoading(false);

		addAlert({
			alertType: Alert.Types.Error,
			message: Liferay.Language.get('error'),
			timeout: false
		});
	};

	const handleSuccess = message => {
		addAlert({
			alertType: Alert.Types.Success,
			message
		});

		setLoading(false);

		refetch();
	};

	return (
		<Card>
			<Card.Body>
				<div className='d-flex justify-content-between align-items-center'>
					<div className='text-secondary'>
						<strong className='font-weight-bold'>
							{Liferay.Language.get('root-endpoint')}
						</strong>

						<span className='ml-1'>
							{window.location.origin + ApisPath}
						</span>
					</div>

					{!tokens.length && (
						<Button
							data-testid='generate-token-button'
							display='primary'
							loading={loading}
							onClick={() => {
								setLoading(true);

								API.apiTokens
									.generate({groupId})
									.then(() => {
										analytics.track('Created API Token');

										handleSuccess(
											Liferay.Language.get(
												'new-token-was-generated'
											)
										);
									})
									.catch(handleError);
							}}
						>
							{Liferay.Language.get('generate-token')}
						</Button>
					)}
				</div>
			</Card.Body>

			{!!tokens.length && (
				<Table
					className='mb-0'
					columns={[
						{
							accessor: 'token',
							cellRenderer: TokenCell,
							label: Liferay.Language.get('token'),
							sortable: false
						},
						{
							accessor: 'lastAccessDate',
							dataFormatter: (val: string) =>
								formatDateToTimeZone(
									val,
									DATE_FORMAT,
									timeZoneId
								),
							label: Liferay.Language.get('last-seen'),
							sortable: false
						},
						{
							accessor: 'createDate',
							dataFormatter: (val: string) =>
								formatDateToTimeZone(
									val,
									DATE_FORMAT,
									timeZoneId
								),
							label: Liferay.Language.get('date-created'),
							sortable: false
						},
						{
							accessor: 'expirationDate',
							dataFormatter: (val: string) =>
								formatDateToTimeZone(val, 'll', timeZoneId),
							label: Liferay.Language.get('expiration'),
							sortable: false
						}
					]}
					items={tokens}
					renderInlineRowActions={({
						data: {expirationDate, token}
					}) => {
						const expired = isExpired(expirationDate);

						return expired ? (
							<Button
								display='primary'
								onClick={() => {
									setLoading(true);

									API.apiTokens
										.revoke({groupId, token})
										.then(() =>
											API.apiTokens.generate({groupId})
										)
										.then(() =>
											handleSuccess(
												Liferay.Language.get(
													'new-token-was-generated'
												)
											)
										)
										.catch(handleError);
								}}
							>
								{Liferay.Language.get('generate-token')}
							</Button>
						) : (
							<>
								<CopyButton text={token} />

								<Button
									onClick={() => {
										open(modalTypes.CONFIRMATION_MODAL, {
											message: (
												<div className='text-secondary'>
													<div>
														<strong>
															{Liferay.Language.get(
																'are-you-sure-you-want-to-revoke-this-token'
															)}
														</strong>
													</div>

													{Liferay.Language.get(
														'you-will-need-to-generate-a-new-token-to-continue-using-this-api'
													)}
												</div>
											),
											modalVariant: 'modal-warning',
											onClose: close,
											onSubmit: () => {
												setLoading(true);

												API.apiTokens
													.revoke({groupId, token})
													.then(() =>
														handleSuccess(
															Liferay.Language.get(
																'token-successfully-revoked'
															)
														)
													)
													.catch(handleError);
											},
											submitButtonDisplay: 'warning',
											title: Liferay.Language.get(
												'revoke-token'
											),
											titleIcon: 'warning-full'
										});
									}}
								>
									{Liferay.Language.get('revoke')}
								</Button>
							</>
						);
					}}
					rowIdentifier='token'
				/>
			)}
		</Card>
	);
};

const ListWithData = compose<any>(
	connector,
	withQuery(
		API.apiTokens.search,
		({groupId}: {groupId: string}) => ({groupId}),
		({data, ...otherParams}) => ({
			tokens: data,
			...otherParams
		})
	),
	withLoading({page: false}),
	withError({page: false})
)(TokenList);

interface IAccessTokenListProps {
	groupId: string;
}

export const AccessTokenList: React.FC<IAccessTokenListProps> = ({groupId}) => (
	<BasePage
		className='access-token-list-root'
		groupId={groupId}
		pageDescription={sub(
			Liferay.Language.get(
				'access-this-workspaces-data-via-api-using-an-access-token.-a-full-list-of-endpoints-is-available-in-the-x'
			),
			[
				<a
					href={URLConstants.APIOverviewDocumentationLink}
					key='API_OVERVIEW_DOCUMENTATION'
					target='_blank'
				>
					{Liferay.Language.get('documentation-fragment')}
				</a>
			],
			false
		)}
		pageTitle={Liferay.Language.get('access-tokens')}
	>
		<ListWithData groupId={groupId} />
	</BasePage>
);

export default withAdminPermission(AccessTokenList);
