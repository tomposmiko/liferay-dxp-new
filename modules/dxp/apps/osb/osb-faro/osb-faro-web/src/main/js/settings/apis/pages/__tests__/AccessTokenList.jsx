import 'test/mock-modal';
import * as API from 'shared/api';
import * as data from 'test/data';
import mockStore from 'test/mock-store';
import Promise from 'metal-promise';
import React from 'react';
import {AccessTokenList} from '../AccessTokenList';
import {cleanup, fireEvent, getByText, render} from '@testing-library/react';
import {getISODate} from 'shared/util/date';
import {mockGetDateNow} from 'test/mock-date';
import {open} from 'shared/actions/modals';
import {Provider} from 'react-redux';
import {StaticRouter} from 'react-router';

jest.unmock('react-dom');

const defaultProps = {
	groupId: '23'
};

const DefaultComponent = props => (
	<StaticRouter>
		<Provider store={mockStore()}>
			<AccessTokenList {...defaultProps} {...props} />
		</Provider>
	</StaticRouter>
);

describe('AccessTokenList', () => {
	beforeAll(() => {
		mockGetDateNow(data.getTimestamp(0));
	});

	afterEach(cleanup);

	it('should render', () => {
		const {container} = render(<DefaultComponent />);

		jest.runAllTimers();

		expect(container).toMatchSnapshot();
	});

	it('should render with a button to generate a token if there are no tokens', () => {
		API.apiTokens.search.mockReturnValueOnce(Promise.resolve([]));

		const {getByTestId} = render(<DefaultComponent />);

		jest.runAllTimers();

		expect(getByTestId('generate-token-button')).toBeTruthy();
	});

	it('should show the generated token in a list and the "Generate Token" button should no longer be visible', () => {
		API.apiTokens.search.mockReturnValueOnce(Promise.resolve([]));

		const {container, getByTestId, queryByTestId} = render(
			<DefaultComponent />
		);

		jest.runAllTimers();

		expect(container.querySelector('.table-root')).toBeNull();
		expect(queryByTestId('generate-token-button')).toBeTruthy();

		fireEvent.click(getByTestId('generate-token-button'));

		jest.runAllTimers();

		expect(container.querySelector('.table-root')).toMatchSnapshot();
		expect(queryByTestId('generate-token-button')).toBeNull();
	});

	it('should open a modal to confirm revoking a token', () => {
		const {getByText} = render(<DefaultComponent />);

		jest.runAllTimers();

		fireEvent.click(getByText('Revoke'));

		jest.runAllTimers();

		expect(open).toBeCalled();
	});

	it('should display the "Generate Token" button in the token row if the token is expired', () => {
		API.apiTokens.search.mockReturnValueOnce(
			Promise.resolve([
				data.mockApiToken({
					expirationDate: getISODate(data.getTimestamp(-1))
				})
			])
		);

		const {container} = render(<DefaultComponent />);

		jest.runAllTimers();

		expect(
			getByText(
				container.querySelector('.row-inline-actions'),
				'Generate Token'
			)
		).toBeTruthy();
	});
});
