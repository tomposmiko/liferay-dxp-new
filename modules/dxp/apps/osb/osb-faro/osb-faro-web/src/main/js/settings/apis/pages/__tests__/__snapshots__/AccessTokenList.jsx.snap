// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AccessTokenList should render 1`] = `
<div>
  <div
    class="settings-base-page-root access-token-list-root"
  >
    <div
      class="maintenance-alert-root"
    />
    <div
      class="content-header"
    >
      <div
        class="header-text"
      >
        <div
          class="d-flex"
        >
          <h3
            class="title-text text-truncate"
          >
            <span
              class="text-truncate"
              data-tooltip="false"
              title=""
            >
              Access Tokens
            </span>
          </h3>
        </div>
        <div
          class="description"
        >
          Access this workspace's data via API using an access token. A full list of endpoints is available in the 
          <a
            href="https://learn.liferay.com/analytics-cloud/latest/en/apis.html"
            target="_blank"
          >
            documentation
          </a>
          .
        </div>
      </div>
      <div
        class="page-actions-container"
      />
    </div>
    <div>
      <div
        class="card card-root"
      >
        <div
          class="card-body"
        >
          <div
            class="d-flex justify-content-between align-items-center"
          >
            <div
              class="text-secondary"
            >
              <strong
                class="font-weight-bold"
              >
                Root Endpoint:
              </strong>
              <span
                class="ml-1"
              >
                http://liferay.com/api/reports
              </span>
            </div>
          </div>
        </div>
        <div
          class="table-responsive table-root flex-grow-1 mb-0"
        >
          <table
            class="table table-autofit table-hover table-heading-nowrap table-nowrap"
          >
            <thead>
              <tr>
                <th
                  class="table-head-title"
                >
                  Token
                </th>
                <th
                  class="table-head-title"
                >
                  Last Seen
                </th>
                <th
                  class="table-head-title"
                >
                  Date Created
                </th>
                <th
                  class="table-head-title"
                >
                  Expiration
                </th>
                <th />
              </tr>
            </thead>
            <tbody
              class="mb-0"
            >
              <tr
                class=""
              >
                <td
                  class=""
                >
                  <span
                    class="text-secondary mr-1"
                  >
                    Token ending in
                  </span>
                  <strong
                    class="font-weight-bold"
                  >
                    4ABC
                  </strong>
                </td>
                <td>
                  Jul 10, 2018
                </td>
                <td>
                  Jul 10, 2018
                </td>
                <td>
                  Aug 9, 2018
                </td>
                <td
                  class="row-inline-actions"
                >
                  <span>
                    <button
                      aria-label="Click to Copy"
                      class="button-root btn btn-secondary"
                      data-clipboard-text="1234ABC"
                      data-tooltip-response="Copied"
                      title="Click to Copy"
                      type="button"
                    >
                      <svg
                        class="icon-root lexicon-icon lexicon-icon-paste"
                        viewBox=""
                      >
                        <use
                          xlink:href="/o/osb-faro-web/dist/sprite.svg#paste"
                        />
                      </svg>
                    </button>
                    <button
                      class="button-root btn btn-secondary"
                      type="button"
                    >
                      Revoke
                    </button>
                  </span>
                </td>
              </tr>
            </tbody>
          </table>
        </div>
      </div>
    </div>
  </div>
</div>
`;

exports[`AccessTokenList should show the generated token in a list and the "Generate Token" button should no longer be visible 1`] = `
<div
  class="table-responsive table-root flex-grow-1 mb-0"
>
  <table
    class="table table-autofit table-hover table-heading-nowrap table-nowrap"
  >
    <thead>
      <tr>
        <th
          class="table-head-title"
        >
          Token
        </th>
        <th
          class="table-head-title"
        >
          Last Seen
        </th>
        <th
          class="table-head-title"
        >
          Date Created
        </th>
        <th
          class="table-head-title"
        >
          Expiration
        </th>
        <th />
      </tr>
    </thead>
    <tbody
      class="mb-0"
    >
      <tr
        class=""
      >
        <td
          class=""
        >
          <span
            class="text-secondary mr-1"
          >
            Token ending in
          </span>
          <strong
            class="font-weight-bold"
          >
            4ABC
          </strong>
        </td>
        <td>
          Jul 10, 2018
        </td>
        <td>
          Jul 10, 2018
        </td>
        <td>
          Aug 9, 2018
        </td>
        <td
          class="row-inline-actions"
        >
          <span>
            <button
              aria-label="Click to Copy"
              class="button-root btn btn-secondary"
              data-clipboard-text="1234ABC"
              data-tooltip-response="Copied"
              title="Click to Copy"
              type="button"
            >
              <svg
                class="icon-root lexicon-icon lexicon-icon-paste"
                viewBox=""
              >
                <use
                  xlink:href="/o/osb-faro-web/dist/sprite.svg#paste"
                />
              </svg>
            </button>
            <button
              class="button-root btn btn-secondary"
              type="button"
            >
              Revoke
            </button>
          </span>
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
